### PR TITLE
Introduce max_results, order_by and page_token arguments to MlflowClient.search_model_versions

### LINKED ISSUE #7417: [FR] Introduce page_token argument to MlflowClient.search_model_versions
### Willingness to contribute

No. I cannot contribute this feature at this time.

### Proposal Summary

Introduce a `page_token` argument to https://mlflow.org/docs/latest/python_api/mlflow.client.html#mlflow.client.MlflowClient.search_model_versions in the MLflow Python client.

### Motivation

> #### What is the use case for this feature? Currently, some MLflow backends (e.g. RestStore) limit the number of results returned by `search_model_versions` requests. Clients must fetch subsequent results by making new requests using a provided page token. Unfortunately, the MLflow Python client's https://mlflow.org/docs/latest/python_api/mlflow.client.html#mlflow.client.MlflowClient.search_model_versions method doesn't accept page tokens, so users can't easily load the next page of results.

> #### Why is this use case valuable to support for MLflow users in general? ^ 

> #### Why is this use case valuable to support for your project(s) or organization? ^ 

> #### Why is it currently difficult to achieve this use case? https://mlflow.org/docs/latest/python_api/mlflow.client.html#mlflow.client.MlflowClient.search_model_versions doesn't accept a page token

### Details

We'll need to introduce a `page_token=None` argument to https://mlflow.org/docs/latest/python_api/mlflow.client.html#mlflow.client.MlflowClient.search_model_versions. We'll also need to introduce `page_token=None` to https://github.com/mlflow/mlflow/blob/7f2152370b8e93ed7bf8cc3e5e1e7ff39cd69ca5/mlflow/store/model_registry/abstract_store.py#L217 and all stores that extend from the Model Registry `AbstractStore`.

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [X] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by serena-ruan**
## Related Issues/PRs

Resolve #7417

## What changes are proposed in this pull request?

Introduce max_results, order_by and page_token arguments to MlflowClient.search_model_versions

## How is this patch tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Introduce max_results, order_by and page_token arguments to MlflowClient.search_model_versions

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by serena-ruan** [comment]
Hi @dbczumar, now for searching model versions I added order by name ASC and version ASC by default, should I order by version DESC instead? Because the larger the number the newer the version is, want to hear your thoughts on this :)

**Comment by WeichenXu123** [comment]
> Hi @dbczumar, now for searching model versions I added order by name ASC and version ASC by default, should I order by version DESC instead? Because the larger the number the newer the version is, want to hear your thoughts on this :)

Shall we order by created date desc ?

**Comment by serena-ruan** [comment]
@dbczumar I'm ready for review :D

File: mlflow/utils/search_utils.py
Comment (harupy) [inline mlflow/utils/search_utils.py:1182]: I filed https://github.com/mlflow/mlflow/pull/7713 to reuse `_Reversor`. Can you update the code once it's merged?

File: mlflow/tracking/client.py
Comment (harupy) [inline mlflow/tracking/client.py:2585]:
```suggestion
        filter_string: Optional[str] = None,
```

File: tests/store/model_registry/test_sqlalchemy_store.py
Comment (harupy) [inline tests/store/model_registry/test_sqlalchemy_store.py:981]:
```suggestion
            MlflowException, match=r"Invalid value for max_results\."
```

File: tests/server/test_handlers.py
Comment (harupy) [inline tests/server/test_handlers.py:623]: `assert_called_once_with` or `assert_called_with` is easier to understand than extracting arguments then asserting it.

https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_once_with

File: mlflow/store/model_registry/sqlalchemy_store.py
Comment (harupy) [inline mlflow/store/model_registry/sqlalchemy_store.py:925]:
```suggestion
                        f"Invalid order by key '{key}' specified. "
                        "Valid keys are "
                        f"'{SearchModelVersionUtils.VALID_ORDER_BY_ATTRIBUTE_KEYS}'",
```

File: tests/tracking/test_model_registry.py
Comment (harupy) [inline tests/tracking/test_model_registry.py:405]: We can try-catch once https://github.com/mlflow/mlflow/pull/7716 is merged.

File: tests/tracking/test_model_registry.py
Comment (harupy) [inline tests/tracking/test_model_registry.py:353]: Separating test cases like below might be easier to understand:

```python
def test_search_model_versions_filter_string():
    ...

def test_search_model_versions_order_by():
    ...

def test_search_model_versions_order_max_results():
    ...

```

File: mlflow/store/model_registry/abstract_store.py
Comment (harupy) [inline mlflow/store/model_registry/abstract_store.py:232]:
```suggestion
                           a ``search_model_versions`` call.
```

File: mlflow/utils/search_utils.py
Comment (BenWilson2) [reply mlflow/utils/search_utils.py:1182]: Merged!

Comment (harupy) [reply mlflow/utils/search_utils.py:1182]: @serena-ruan Let's rebase on master and reuse `_Reversor`!

File: mlflow/store/model_registry/rest_store.py
Comment (BenWilson2) [inline mlflow/store/model_registry/rest_store.py:331]: similar question as above regarding providing a default orderby in the paginated request if none is specified

File: tests/tracking/test_model_registry.py
Comment (BenWilson2) [reply tests/tracking/test_model_registry.py:280]: +1 that test parametrize is pretty dense.

File: mlflow/store/model_registry/file_store.py
Comment (serena-ruan) [reply mlflow/store/model_registry/file_store.py:731]: If no condition is specified and to avoid a tie condition, I explicitly added sorting by last_updated_timestamp DESC in search_utils.py.

File: mlflow/store/model_registry/sqlalchemy_store.py
Comment (serena-ruan) [reply mlflow/store/model_registry/sqlalchemy_store.py:927]: Good idea! I'll update

Comment (serena-ruan) [reply mlflow/store/model_registry/sqlalchemy_store.py:927]: Actually the behavior for all order_by related functions are raising errors immediately without recording invalid keys, and the error for this function specifically is caught inside 'SearchModelVersionUtils.parse_order_by_for_search_model_versions' which just analyzes one single order_by_clause a time. I agree with you that we should record invalid keys and raise them all in once, but could I update all of them together in a follow-up PR? This PR is quite long enough, and the change would affect other components like experiments as well.

File: tests/tracking/test_model_registry.py
Comment (serena-ruan) [reply tests/tracking/test_model_registry.py]: Done.

File: mlflow/store/model_registry/file_store.py
Comment (BenWilson2) [reply mlflow/store/model_registry/file_store.py:731]: ah yes :) The tie breaker. Gotcha!

File: mlflow/store/model_registry/sqlalchemy_store.py
Comment (BenWilson2) [reply mlflow/store/model_registry/sqlalchemy_store.py:927]: Agreed that updating this logic is out of scope for this PR. I think a follow up PR to make  a more pleasant UX for this would be great :)

**Comment by serena-ruan** [comment]
@BenWilson2 @harupy Could you help me rerun the cancelled tasks? And if it passes I'm waiting for approval :D

**Comment by BenWilson2** [comment]
@serena-ruan the tasks were cancelled due to a timeout from pip resolution tree logic. Could you merge master branch into your PR (that should fix it :) )?

**Comment by serena-ruan** [comment]
Hi @dbczumar, @WeichenXu123 do you have any more comments? (Or approval :D)

**Review by BenWilson2 — APPROVED** [review]
LGTM! Great work @serena-ruan !

File: mlflow/protos/model_registry.proto
Comment (harupy) [inline mlflow/protos/model_registry.proto:533]: Curious why we made this change

File: mlflow/utils/search_utils.py
Comment (harupy) [inline mlflow/utils/search_utils.py:1139]:
```suggestion
        model_versions = [
            for mv in model_versions if mv.current_stage != STAGE_DELETED_INTERNAL
        ]
```

to avoid returning a filter object on line 1141.

File: tests/store/model_registry/test_file_store.py
Comment (harupy) [inline tests/store/model_registry/test_file_store.py:1431]:
```suggestion
        assert len(mv1) == 2
        assert mv1[0].name == "model1"
```

Assertion on multiple conditions makes debugging difficlt. It should be broken down.

File: tests/tracking/test_model_registry.py
Comment (harupy) [inline tests/tracking/test_model_registry.py:267]: It's a bit challenging for me to understand what's going on in this test. Can we remove parametrization? Do we need to check pagination?

Comment (harupy) [reply tests/tracking/test_model_registry.py:267]: I'll update this test in a follow-up.

File: mlflow/utils/search_utils.py
Comment (harupy) [inline mlflow/utils/search_utils.py:1172]: Can `last_updated_timestamp` really break ties? It's possible that two model versions are created within 1 millisecond. Can we use `name` and `version` instead?

File: tests/store/model_registry/test_file_store.py
Comment (harupy) [inline tests/store/model_registry/test_file_store.py:775]:
```suggestion
    assert len(mvds) == 1
```

The value to check should be on the left-hand side.

File: tests/store/model_registry/test_file_store.py
Comment (harupy) [inline tests/store/model_registry/test_file_store.py:785]: Do we need this prefix?

File: mlflow/protos/model_registry.proto
Comment (serena-ruan) [reply mlflow/protos/model_registry.proto:533]: I thought the previous default value is set to 200k just as a placeholder? I followed the same default value as registered models.

File: mlflow/utils/search_utils.py
Comment (serena-ruan) [reply mlflow/utils/search_utils.py:1172]: I'm trying to keep the same behavior as databricks backend, Weichen mentioned databricks backend orders by last_updated_timestamp DESC by default, and how does it deal with such scenario? By name ASC + version DESC?

File: tests/store/model_registry/test_file_store.py
Comment (serena-ruan) [reply tests/store/model_registry/test_file_store.py:785]: Not really, I just use it as a convention as other tests :) Do you want me to remove it?

Comment (harupy) [reply tests/store/model_registry/test_file_store.py:785]: Let's remove it!

File: mlflow/utils/search_utils.py
Comment (harupy) [reply mlflow/utils/search_utils.py:1170]: Let me check.

Comment (harupy) [reply mlflow/utils/search_utils.py:1170]: The default ordering is different than a tiebreaker.

Comment (serena-ruan) [reply mlflow/utils/search_utils.py:1170]: I see, I should update the default to be ["last_updated_timestamp DESC", "name ASC", "version DESC"]

**Comment by harupy** [comment]
@serena-ruan Let me test this on Databricks.

File: mlflow/utils/search_utils.py
Comment (harupy) [inline mlflow/utils/search_utils.py:1091]:
```suggestion
    NUMERIC_ATTRIBUTES = {"version", "creation_timestamp", "last_updated_timestamp"}
```

Can we change `version_number` to `version`?

File: mlflow/protos/model_registry.proto
Comment (harupy) [reply mlflow/protos/model_registry.proto:533]: How about 1000 (default) and 10000 (max)?

**Comment by harupy** [comment]
I tested on Databricks and it worked :)

**Review by harupy — APPROVED** [review]
Left a few more comments! LGTM once they are addressed :) Thanks for the hard work!

File: mlflow/utils/search_utils.py
Comment (serena-ruan) [reply mlflow/utils/search_utils.py:1091]: Actually no, because version is a keyword in sqlparse, I map it to version_number basically the same reason as 'source_path' versus 'source'.

**Comment by serena-ruan** [comment]
@harupy Should I update this newly added 'mlflow/store/_unity_catalog/registry/rest_store.py' to follow the same function pattern of search_model_versions? The proto definition doesn't include order_by in it as for now.

File: mlflow/protos/model_registry.proto
Comment (dbczumar) [inline mlflow/protos/model_registry.proto:533]: @serena-ruan @harupy Can we revert this change to avoid making backwards-incompatible changes to the MLflow REST API? Can we instead add the following sentence to the docstring: "Backends may choose a lower default value and maximum threshold." ?

**Review by dbczumar — APPROVED** [review]
Awesome stuff, thanks @serena-ruan ! LGTM once https://github.com/mlflow/mlflow/pull/7623/files#r1112501352 is addressed (small reversion to avoid backwards-incompatible REST API changes).

File: mlflow/protos/model_registry.proto
Comment (harupy) [reply mlflow/protos/model_registry.proto]: @dbczumar Reverted the change and added the sentence 👍

**Comment by serena-ruan** [comment]
> @harupy Should I update this newly added 'mlflow/store/_unity_catalog/registry/rest_store.py' to follow the same function pattern of search_model_versions? The proto definition doesn't include order_by in it as for now.

The failing lint test is for this newly added `mlflow/store/_unity_catalog/registry/rest_store.py`, should we fix it or ignore pylint here?

**Comment by harupy** [comment]
@serena-ruan I pushed a commit to fix the failed checks.

File: mlflow/utils/search_utils.py
Comment (harupy) [reply mlflow/utils/search_utils.py:1091]: Got it. We could implement our own parser to avoid this issue.

**Pull request merged**
